python/ql/lib/semmle/python/frameworks/Bottle.qll

```diff
@@ -0,0 +1,179 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `bottle` PyPI package.
+ * See https://bottlepy.org/docs/dev/.
+ */
+
+private import python
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+private import semmle.python.dataflow.new.RemoteFlowSources
+private import semmle.python.frameworks.internal.InstanceTaintStepsHelper
+private import semmle.python.frameworks.Stdlib
+
+/**
+ * INTERNAL: Do not use.
+ *
+ * Provides models for the `bottle` PyPI package.
+ * See https://bottlepy.org/docs/dev/.
+ */
+module Bottle {
+  /** Gets a reference to the `bottle` module. */
+  API::Node bottle() { result = API::moduleImport("bottle") }
+
+  /** Provides models for the `bottle` module. */
+  module BottleModule {
+    /**
+     * Provides models for Bottle applications.
+     */
+    module App {
+      /** Gets class `bottle.Bottle`) */
+      API::Node cls() { result = API::moduleImport("bottle").getMember("Bottle") }
+
+      /** Gets a reference to a Bottle application (an instance of `bottle.Bottle`) */
+      API::Node instance() { result = cls().getReturn() }
+
+      /** Gets a reference to a Bottle application (an instance of `bottle.app`) */
+      API::Node app() { result = bottle().getMember("app").getReturn() }
+    }
+
+    /** Provides models for functions that are possible "views" */
+    module View {
+      /**
+       * A Bottle view callable, that handles incoming requests.
+       */
+      class ViewCallable extends Function {
+        ViewCallable() { this = any(BottleRouteSetup rs).getARequestHandler() }
+      }
+
+      private class BottleRouteSetup extends Http::Server::RouteSetup::Range, DataFlow::CallCfgNode {
+        BottleRouteSetup() {
+          this =
+            [
+              App::instance()
+                  .getMember(["route", "get", "post", "put", "delete", "patch"])
+                  .getACall(),
+              App::app().getMember(["route", "get", "post", "put", "delete", "patch"]).getACall(),
+              bottle().getMember(["route", "get", "post", "put", "delete", "patch"]).getACall()
```

Comment:
It seems the list of calls are the same in all situations by design of the bottle framework, so we could have a string predicate for it. That would ensure that we do not forget one function in one of the cases.
Alternatively, you could factor out the `API::Node`, like `exists(API::Node app | app in [App::instance(), App::app(), bottle() | this = app..getMember(["route", "get", "post", "put", "delete", "patch"])
                  .getACall() )`